diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/WhenMustHaveElseRule.kt

```diff
@@ -0,0 +1,82 @@
+package org.cqfn.diktat.ruleset.rules
+
+import com.pinterest.ktlint.core.KtLint
+import com.pinterest.ktlint.core.Rule
+import com.pinterest.ktlint.core.ast.ElementType
+import org.cqfn.diktat.common.config.rules.RulesConfig
+import org.cqfn.diktat.ruleset.constants.Warnings
+import org.cqfn.diktat.ruleset.utils.findAllNodesWithSpecificType
+import org.cqfn.diktat.ruleset.utils.hasChildOfType
+import org.cqfn.diktat.ruleset.utils.prettyPrint
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.CompositeElement
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.LeafPsiElement
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.PsiWhiteSpaceImpl
+import org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.java.PsiBlockStatementImpl
+import java.lang.StringBuilder
+
+/**
+ * Rule 3.10: 'when' statement must have else branch, unless when condition variable is enumerated or sealed type
+ *
+ * Current limitations and FixMe:
+ * If a when statement of type enum or sealed contains all values of a enum - there is no need to have "else" branch.
+ * The compiler can issue a warning when it is missing.
+ */
+@Suppress("ForbiddenComment")
+class WhenMustHaveElseRule : Rule("no-else-in-when") {
+
+    private lateinit var configRules: List<RulesConfig>
+    private lateinit var emitWarn: ((offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit)
+    private var isFixMode: Boolean = false
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       params: KtLint.Params,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        configRules = params.getDiktatConfigRules()
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == ElementType.WHEN) {
+            checkEntries(node)
+        }
+    }
+
+
+    //FixMe: If a when statement of type enum or sealed contains all values of a enum - there is no need to have "else" branch.
+
+    private fun checkEntries(node: ASTNode) {
+        if (!checkElses(node)) {
+            Warnings.WHEN_WITHOUT_ELSE.warnAndFix(configRules, emitWarn, isFixMode, "else was not found", node.startOffset) {
+                val whenEntryElse = CompositeElement(ElementType.WHEN_ENTRY)
+                node.addChild(whenEntryElse, node.lastChildNode.treePrev)
+                addChildren(whenEntryElse)
+                node.addChild(PsiWhiteSpaceImpl("\n"), whenEntryElse)
+            }
+        }
+    }
+
+    private fun checkElses(node: ASTNode): Boolean = node.findAllNodesWithSpecificType(ElementType.WHEN_ENTRY).any {
+            it.hasChildOfType(ElementType.ELSE_KEYWORD)
+        }
```

Comment:
```suggestion
    private fun checkElses(node: ASTNode): Boolean = (node.psi as KtWhenExpression).elseExpression != null
```
Either use psi where some of operations are already implemented, or you should use `getAllChildrenWithType` for appropriate node instead of `findAllNodesWithSpecificType`, which does full DFS and is more suited for cases when tree structure can be complex.